Desktop_Interface/functiongencontrol.cpp

```diff
@@ -86,65 +83,103 @@ void functionGenControl::waveformName(QString newName, ChannelData& channel, int
     QByteArray remainingData = fptr.readAll();
     char *dataString = remainingData.data();
 
-    free(channel.samples);
-    channel.samples = (unsigned char *) malloc(channel.length);
+	channel.samples.resize(channel.length);
 
     int dummy;
     char *dataStringCurrent = dataString;
-    for (int i=0;i<channel.length;i++){
+    for (int i = 0; i < channel.length; i++)
+	{
         sscanf(dataStringCurrent, "%d", &dummy);
         dataStringCurrent += strcspn(dataStringCurrent, "\t") + 1;
-        channel.samples[i] = (unsigned char) dummy;
+        channel.samples[i] = uint8_t(dummy);
     }
 
 #else
-    QString directory = QCoreApplication::applicationDirPath();
 
-    directory.append("/waveforms/");
-    directory.append(newName);
-    QByteArray temp = encode(directory);
-    char *fileName = temp.data();
+    QByteArray filePath = QCoreApplication::applicationDirPath()
+		.append("/waveforms/").append(newName).toUtf8();
```

Comment:
I would be wary about passing a UTF-8 array to `fopen()`.  It may work, but I wouldn't trust it to be cross-platform and safe in all locales.  Local 8-bit should be much safer.